src/Mod/CAM/PathSimulator/AppGL/MillPathSegment.cpp

```diff
@@ -0,0 +1,242 @@
+#define PX 0
+#define PY 1
+#define PZ 2
+
+
+namespace MillSim
+{
+
+bool IsVerticalMotion(MillMotion* m1, MillMotion* m2)
+{
+    return (m1->z != m2->z && EQ_FLOAT(m1->x, m2->x) && EQ_FLOAT(m1->y, m2->y));
+}
+
+bool IsArcMotion(MillMotion* m)
+{
+    if (m->cmd != eRotateCCW && m->cmd != eRotateCW) {
+        return false;
+    }
+    return fabs(m->i > EPSILON) || fabs(m->j) > EPSILON;
+}
+
+float MillPathSegment::mResolution = 1;
+float MillPathSegment::mSmallRadStep = (PI / 8);
+
+MillPathSegment::MillPathSegment(EndMill* _endmill, MillMotion* from, MillMotion* to)
+    : mShearMat {1.0f,
+                 0.0f,
+                 0.0f,
+                 0.0f,
+                 0.0f,
+                 1.0f,
+                 0.0f,
+                 0.0f,
+                 0.0f,
+                 0.0f,
+                 1.0f,
+                 0.0f,
+                 0.0f,
+                 0.0f,
+                 0.0f,
+                 1.0f}
+{
+
+    MotionPosToVec(mStartPos, from);
+    MotionPosToVec(mDiff, to);
+    vec3_sub(mDiff, mDiff, mStartPos);
+    mXYDistance = sqrtf(mDiff[PX] * mDiff[PX] + mDiff[PY] * mDiff[PY]);
+    mZDistance = fabsf(mDiff[PY]);
+    mXYZDistance = sqrtf(mXYDistance * mXYDistance + mDiff[PZ] * mDiff[PZ]);
+    mXYAngle = atan2f(mDiff[PY], mDiff[PX]);
+    endmill = _endmill;
+    mStartAngRad = mStepAngRad = 0;
+    if (IsArcMotion(to)) {
+        mMotionType = MTCurved;
+        mRadius = sqrtf(to->j * to->j + to->i * to->i);
+        mSmallRad = mRadius <= endmill->radius;
+
+        if (mSmallRad) {
+            mStepAngRad = mSmallRadStep;
+        }
+        else {
+            mStepAngRad = asinf(mResolution / mRadius);
+            if (mStepAngRad > MAX_SEG_DEG) {
+                mStepAngRad = MAX_SEG_DEG;
+            }
+            else if (mStepAngRad < NIN_SEG_DEG) {
+                mStepAngRad = NIN_SEG_DEG;
+            }
+        }
+
+        MotionPosToVec(mCenter, from);
+        mCenter[PX] += to->i;
+        mCenter[PY] += to->j;
+        mArcDir = to->cmd == eRotateCCW ? -1.f : 1.f;
+        mStartAngRad = atan2f(mCenter[PX] - from->x, from->y - mCenter[PY]);
+        float endAng = atan2f(mCenter[PX] - to->x, to->y - mCenter[PY]);
+        mSweepAng = (mStartAngRad - endAng) * mArcDir;
+        if (mSweepAng < EPSILON) {
+            mSweepAng += PI * 2;
+        }
+        numSimSteps = (int)(mSweepAng / mStepAngRad) + 1;
+        mStepAngRad = mArcDir * mSweepAng / numSimSteps;
+        if (mSmallRad) {
+            // when the radius is too small, we just use the tool itself to carve the stock
+            mShape = endmill->toolShape;
+        }
+        else {
+            endmill->GenerateArcSegmentDL(mRadius,
+                                           mStepAngRad * SWEEP_ARC_PAD,
+                                           mDiff[PZ] / numSimSteps,
+                                           &mShape);
+            numSimSteps++;
+        }
+
+        isMultyPart = true;
```

Comment:
This is very big block of code - probably could be it's own method if so, the else body could also be extracted to be method.

Reply:
ok. for next time.